rfcs/2023-12-25-proxy-implementation.md

```diff
@@ -0,0 +1,108 @@
+# SMP proxy implementation concerns
+
+* SMP proxy traffic should be indistinguishable from ordinary relay operation.
+* Proxy servers should be implemented by the same binary to prevent maintenance bloat.
+
+This leads to the proxy functionality added directly to the SMP protocol, raising its version.
+
+## Involved parties
+
+### Client
+
+> Client, from the servers' point of view, means the agent running inside a client app. An app/agent protocol is unchanged.
+
+A client that wants to utilize proxy functionality should request protocol version no lower than adds support for it.
+
+A client should be able to configure its usage of SMP proxies.
+Due to proxy/relay integration, there can be some usage patterns (or some blend of them):
+
+1. Single connection for everything. A client uses its only connection for both creating queues for reception and forwarding messages to remote relays. This can be favorable for power-constrained operation or utilizing a TOR hidden service.
```

Comment:
should use the same "session" scope approach - per user or per entity. No reason to split proxy and the rest of the traffic if they are in the same protocol.